research/picai/prepare_data.py

```diff
@@ -0,0 +1,209 @@
+import argparse
+import json
+import os
+from pathlib import Path
+from typing import Optional, Sequence, Tuple
+
+import numpy as np
+import SimpleITK as sitk
+from preprocessing import (
+    AlignOriginAndDirection,
+    BinarizeAnnotation,
+    CentreCropAndOrPad,
+    PicaiCase,
+    PreprocessingSettings,
+    Resample,
+    ResampleToFirstScan,
+    preprocess,
+)
+
+DEFAULT_TRANSFORMS = [
+    Resample(),
+    CentreCropAndOrPad(),
+    ResampleToFirstScan(),
+    AlignOriginAndDirection(),
+    BinarizeAnnotation(),
+]
+
+
+def get_labels(paths_for_each_sample: Sequence[Tuple[Sequence[Path], Path]]) -> Sequence[float]:
+    _, annotation_paths = zip(*paths_for_each_sample)
+    case_labels = []
+    for path in annotation_paths:
+        annotation = sitk.ReadImage(path)
+        annotation_array = sitk.GetArrayFromImage(annotation)
+        case_label = float(np.max(annotation_array))
+        case_labels.append(case_label)
+    return case_labels
+
+
+def generate_dataset_json(
+    paths_for_each_sample: Sequence[Tuple[Sequence[Path], Path]],
+    write_dir: Path,
+    splits_path: Optional[Path] = None,
+) -> None:
+
+    labels = get_labels(paths_for_each_sample)
+
+    # Since we are storing paths inside a json file, we need to convert to strings to be serializable.
+    # Also pack in label for each samples in the tuple containing scan paths and annotation path.
+    scan_annotation_label_list = [
+        (list(map(str, tup[0])), str(tup[1]), lbl) for tup, lbl in zip(paths_for_each_sample, labels)
+    ]
+
+    if splits_path is None:
+        # If splits_path is None, create a singe dataset overview
+        d = {}
+        d["image_paths"], d["label_paths"], d["case_label"] = zip(*scan_annotation_label_list)
+        write_path = os.path.join(write_dir, "train-fold-all.json")
+        with open(write_path, "w") as f:
+            json.dump(d, f)
+    else:
+        # If splits_path is not None, create a dataset overview for each split
+        with open(splits_path, "r") as splits_f:
+            splits = json.load(splits_f)
+
+        for i, split in enumerate(splits):
+            # Add image (scan) and label (annotation) paths that include any subject ids from split
+            # For both train and val
+            train_d, val_d = {}, {}
+            train_list = [
+                tup
+                for tup in scan_annotation_label_list
```

Comment:
Breaking this tuple apart here as `(_, label_path, _)` or something would be a bit more readable as to what you're testing for.

Reply:
It also might be worth breaking this filtration into a helper function for readability since you're essentially doing the same operation twice.